Nuvoton: Fix FPGA CI test failing

ccli8 opened the pull request:
### Description

This PR is to pass FPGA CI test shield tests on Nuvoton targets. It has the following major bugfix:

- Skip some Nuvoton targets not supporting pull-up/pull-down input mode
- Force enum `PinName` to 32-bit required to encode module binding information in it
- Exclude USB UART from testing which is dedicated to USB VCOM
- Fix IP initialization sequence
- Fix I2C NACK error
- Fix redundant SPI clock generation
- Fix redundant call to UART IRQ handler
- Support GPIO input pull-high/pull-low (not all Nuvoton targets support it)
- Other target-specific bugfix


### Pull request type

    [X] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

ccli8 committed:
[Nuvoton] Exclude USB UART from testing

USB UART is dedicated to USB COM and so must exclude from FPGA CI testing.

ccli8 committed:
[Nuvoton] Fix I2C NACK error

Fix logic error on replying NACK at the end of transfer.

This is also to fix FPGA CI test mbed_hal_fpga_ci_test_shield-i2c/
i2c - test single byte read i2c API.

ccli8 committed:
[Nuvoton] Fix redundant SPI clock generation

Fix SPI clocks are generated redundantly at the end of transfer.

This is also to fix FPGA CI test mbed_hal_fpga_ci_test_shield-spi/
SPI - async mode.

ccli8 committed:
[NANO130] Remove SPI MOSI1/MISO1 pins from pinmap

MOSI1/MISO1 are used in second bit of 2-bit transfer mode and cannot be used
for normal MOSI/MISO. Remove them from pinmap.

This is also to fix FPGA CI test mbed_hal_fpga_ci_test_shield-spi/
SPI - basic test.

ccli8 committed:
[M2351] Fix CLK_SetModuleClock_S(...) error with SPI

Fix SPI module index error in modidx_ns_tab table in CLK_SetModuleClock_S().
Need to update secure image for this bugfix.

This is also to fix FPGA CI test mbed_hal_fpga_ci_test_shield-spi/
SPI - init/free test all pins.

ccli8 committed:
[Nuvoton] Force enum PinName to 32-bit

NU_PINNAME_BIND(...) requires enum PinName to be 32-bit to encode module
binding information in it.

ccli8 committed:
[M2351] Fix pinmap table error with SPI clock pin

ccli8 committed:
[Nuvoton] Fix IP initialization sequence

Better IP initialization sequence:
1. Configure IP pins
2. Select IP clock source and then enable it
3. Reset the IP (SYS_ResetModule)

NOTE1: IP reset takes effect regardless of IP clock. So it doesn't matter if
       IP clock enable is before IP reset.
NOTE2: Non-configured pins may disturb IP's state, so IP pinout first and then
       IP reset.
NOTE3: IP reset at the end of IP initialization sequence can cover unexpected
       situation.

ccli8 committed:
[M453] Classify by M45xD/M45xC and M45xG/M45xE

M451 series can classify by M45xD/M45xC and M45xG/M45xE. To support this
classification:
1.  Create TARGET_M45xD_M45xC and TARGET_M45xG_M45xE targets.
2.  Mark NUMAKER_PFM_M453 belongs to TARGET_M45xG_M45xE by 'extra_labels_add'
    in targets.json.
3.  Fix pin name table according to the classification.
4.  Fix pinmap table according to the classification.

ccli8 committed:
[Nuvoton] Fix redundant call to UART IRQ handler

Honor RxIrq/TxIrq to avoid redundant call to UART IRQ handler.

This is also to fix FPGA CI test mbed_hal_fpga_ci_test_shield-uart.

ccli8 committed:
[Nuvoton] Support GPIO input pull-high/pull-low

In Nuvoton, only new-design chips support GPIO input pull-high/pull-low modes.
Targets not supporting this feature are listed below:

- NUMAKER_PFM_NANO130
- NUMAKER_PFM_NUC472
- NUMAKER_PFM_M453

ccli8 committed:
FPGA: Free up I2C pins to avoid interference with following I2C tests

The most suitable place to free up I2C pins is in i2c_free(). Due to
i2c_free() not available in I2C HAL, we free up I2C pins manually by
configuring them back to GPIO.

Without free-up of I2C pins, SDA/SCL pins of the same I2C peripheral may
share by multiple ports due to 'all ports' tests here, and the following I2C
tests would be subject to interference by shared ports.

ccli8 committed:
[Nuvoton] Free up peripheral pins in peripheral free-up HAL API

Without free-up of peripheral pins, peripheral pins of the same peripheral may
share by multiple ports after port iteration, and this peripheral may fail with
pin interference.

ccli8 committed:
[M487] Exclude A2/A3 from testing for NuMaker-IoT-M487 V1.3

Since NuMaker-IoT-M487 V1.3, A2/A3 are dedicated to on-board ESP8266 WiFi
module RTS/CTS pins and so must exclude from FPGA CI testing.

0xc0170 commented on TESTS/mbed_hal_fpga_ci_test_shield/gpio/main.cpp:
```diff
@@ -71,7 +71,12 @@ void test_basic_input_output(PinName pin)
+    // Some targets don't support pull-up/pull-down mode.
+#if !defined(TARGET_NANO100) &&         \
+    !defined(TARGET_NUC472) &&          \
+    !defined(TARGET_M451)
```
How can we eliminate these if target. Is this functionality a must ? Can we check differently if this mode is supported?

Having this ifdef is not ideal, means we are testing something is not either supported by all targets or we are missing a check to ask for "is this supported, test it".

ccli8 replied:
Check `gpio_mode(...)` and `pin_mode(...)` but find no spec on the `mode` parameter. It seems most targets support `PullDown`/`PullUp` input pull mode. However, old-chip based Nuvoton targets like above don't support it. And there's no configuration option to disable it.

```
void gpio_mode(gpio_t *obj, PinMode mode);
void pin_mode(PinName pin, PinMode mode);
```

0xc0170 replied:
@ARMmbed/mbed-os-hal Does this look correct as it is?

fkjagodzinski replied:
@ccli8 The whole `// Test input, pull-up mode.` should be excluded if these targets do not support pull-up mode.

fkjagodzinski replied:
> How can we eliminate these if target. Is this functionality a must ? Can we check differently if this mode is supported?
> 
> Having this ifdef is not ideal, means we are testing something is not either supported by all targets or we are missing a check to ask for "is this supported, test it".

Frankly, I didn't expect any target that does not support pull-up/pull-down modes. The only thing that comes to my mind, to make this a bit more readable, is to define a new macro in *targets.json* file. That could be either `"GPIO_PINMODE"` in `"device_has"` section or `"GPIO_PINMODE_NOT_SUPPORTED"` in `"macros"`. @ccli8, @0xc0170, @OPpuolitaival, what do you think?

ccli8 replied:
Before that, I think the `mode` parameter should be well defined. It just allows `PullNone`/`PullUp`/`PullDown`, or it can also allow other modes like target-specific? Only meaningful for input mode?

```C
void gpio_mode(gpio_t *obj, PinMode mode);
void pin_mode(PinName pin, PinMode mode);
```

fkjagodzinski commented on TESTS/mbed_hal_fpga_ci_test_shield/gpio/main.cpp:
```diff
@@ -80,13 +85,23 @@ void test_basic_input_output(PinName pin)
     gpio_mode(&gpio, PullDown);
     tester.gpio_write(MbedTester::LogicalPinGPIO0, 0, false);
     wait_us(HI_Z_READ_DELAY_US);
+    // Some targets don't support pull-up/pull-down mode.
```
Same as above -- the whole `// Test input, pull-down mode.` should be excluded if pull-down is not supported.

ccli8 replied:
@fkjagodzinski Including above, fixed as commented

fkjagodzinski replied:
Thanks for a quick update. Don't you think the `#if` statement is a bit too wide now? I thought that only the pull-up and pull-down modes are unsupported; pull-none was OK, wasn't it? I'd rather see [these lines](https://github.com/ARMmbed/mbed-os/blob/398515a3678f78847118242a536de3ba0b058035/TESTS/mbed_hal_fpga_ci_test_shield/gpio/main.cpp#L70-L98) enclosed in `#if !defined(TARGET_NANO100) ...`. I do not have these targets available for local tests, so will let you decide. :wink:

ccli8 replied:
@fkjagodzinski Chip may have fixed (non-configurable) input pull mode on these targets. To be not confusing, pull modes (`PullNone`/`PullUp`/`PullDown`) are not supported on these targets.

fkjagodzinski replied:
In that case, we can leave the `#if` statements as they are.

fkjagodzinski commented on TESTS/mbed_hal_fpga_ci_test_shield/gpio/main.cpp:
```diff
@@ -142,31 +162,51 @@ void test_explicit_input(PinName pin)
+    // Some targets don't support pull-up/pull-down mode.
+#if !defined(TARGET_NANO100) &&         \
+    !defined(TARGET_NUC472) &&          \
+    !defined(TARGET_M451)
```
Since every step of this test (`test_explicit_input()`) sets one of the `PinModes`, the whole test should be skipped if not supported. I think the `#if` statement should be moved to the declaration of `cases` below.

ccli8 replied:
@fkjagodzinski Fixed as commented

ccli8 committed:
FPGA: Skip some Nuvoton targets not supporting input pull-up/pull-down mode

The Nuvoton targets listed below don't support input pull-up/pull-down mode
and so are skipped for test:
- NUMAKER_PFM_NANO130
- NUMAKER_PFM_NUC472
- NUMAKER_PFM_M453

0xc0170 reviewed (approved):
`TESTS/mbed_hal_fpga_ci_test_shield/gpio/main.cpp` change still needs to be approved by @fkjagodzinski
